src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+        sage: G_I = matrix.identity(QQ,2)
+        sage: H_J = 2*G_I
+        sage: check_gevp_feasibility(0,xi,eta,G_I,None,H_J,None,0)
+        (False, (0, 0), (0, 0))
+
+    When `\cos\theta` is zero, the inequality (42) in Theorem 7.3
+    [SS2016]_ is just an inner product with `v` which we can make
+    positive by ensuring that all of the entries of `H_{J}` are
+    positive. So, if any of the rows of ``G_I_c_T`` contain a negative
+    entry, (42) will fail::
+
+        sage: from sage.geometry.cone_critical_angles import(
+        ....:   check_gevp_feasibility)
+        sage: xi = vector(QQ, [1/2,1/2,1/2,1/2])
+        sage: eta = xi
+        sage: G_I = matrix.identity(QQ,4)
+        sage: G_I_c_T = matrix(QQ, [[0,-1,0,0]])
+        sage: H_J = G_I
+        sage: check_gevp_feasibility(0,xi,eta,G_I,G_I_c_T,H_J,None,0)
+        (False, (0, 0, 0, 0), (0, 0, 0, 0))
+
+    Likewise we can make (43) fail in exactly the same way::
+
+        sage: from sage.geometry.cone_critical_angles import(
+        ....:   check_gevp_feasibility)
+        sage: xi = vector(QQ, [1/2,1/2,1/2,1/2])
+        sage: eta = xi
+        sage: G_I = matrix.identity(QQ,4)
+        sage: G_I_c_T = matrix(QQ, [[0,1,0,0]])
+        sage: H_J = G_I
+        sage: H_J_c_T = matrix(QQ, [[0,-1,0,0]])
+        sage: check_gevp_feasibility(0,xi,eta,G_I,G_I_c_T,H_J,H_J_c_T,0)
+        (False, (0, 0, 0, 0), (0, 0, 0, 0))
+
+    Finally, if we ensure that everything works, we get back a feasible
+    result along with the vectors (scaled `\xi` and `\eta`) that worked::
+
+        sage: from sage.geometry.cone_critical_angles import(
+        ....:   check_gevp_feasibility)
+        sage: xi = vector(QQ, [1/2,1/2,1/2,1/2])
+        sage: eta = xi
+        sage: G_I = matrix.identity(QQ,4)
+        sage: G_I_c_T = matrix(QQ, [[0,1,0,0]])
+        sage: H_J = G_I
+        sage: H_J_c_T = matrix(QQ, [[0,1,0,0]])
+        sage: check_gevp_feasibility(0,xi,eta,G_I,G_I_c_T,H_J,H_J_c_T,0)
+        (True, (1/2, 1/2, 1/2, 1/2), (1/2, 1/2, 1/2, 1/2))
+    """
+    infeasible_result = (False, 0*xi, 0*eta)
+    if min(xi) <= -epsilon or min(eta) <= -epsilon:
+        # xi or eta isn't in the interior of the nonnegative orthant,
+        # so skip this (non-)solution.
+        return infeasible_result
+
+    # Rescale xi to satisfy (44), and rescale eta by the same amount,
+    # because (xi,eta) needs to remain in the same one-dimensional
+    # eigenspace.
+    scale = ~((G_I*xi).norm())
+    xi_hat = xi*scale
+    eta_hat = eta*scale
+
+    # Now check that (45) is satisfied.
+    if ((H_J*eta_hat).norm() - 1).abs() > epsilon:
+        return infeasible_result
+
+    # And check that (42,43) are satisfied.
+    v = H_J*eta_hat
+    rhs = v - cos_theta*G_I*xi_hat
+
+    if any( x < -epsilon for x in G_I_c_T*rhs ):
+        return infeasible_result
+
+    u = G_I*xi_hat
+    rhs = u - cos_theta*H_J*eta_hat
+    if any( x < -epsilon for x in H_J_c_T*rhs):
+        return infeasible_result
```

Comment:
```suggestion
    v = H_J * eta_hat
    rhs = v - cos_theta * G_I * xi_hat

    if any(x < -epsilon for x in G_I_c_T * rhs):
        return infeasible_result

    u = G_I * xi_hat
    rhs = u - cos_theta * H_J * eta_hat
    if any(x < -epsilon for x in H_J_c_T * rhs):
        return infeasible_result
```
PEP8

Reply:
I fixed some of these, but PEP8 actually recommends writing products without spaces when addition/subtraction (or any lower-precedence operation) are involved: https://peps.python.org/pep-0008/#other-recommendations

As a result, `v - cos_theta*G_I*xi_hat` is probably correct -- it's one of the examples they give. The expression,

```python
x < -epsilon for x in H_J_c_T*rhs
```

is more ambiguous, but in theory you could be multiplying the result of the list comprehension by `rhs`. Writing `H_J_c_T*rhs` with no spaces helps indicate (to me, anyway) that it's one quantity and binds stronger than the `for foo in bar`. I'm still happy to change it if you care, though.

Reply:
> As a result, `v - cos_theta*G_I*xi_hat` is probably correct

I strongly agree. It's correct per PEP 8 and much more readable than the fully "spaced out" version.

> Writing `H_J_c_T*rhs` with no spaces

This one is trickier to justify per PEP 8, but I still support it; not with the explanation that you give, but rather because these terms (products) appear right next to polynomial expression where the monomials are written in this way without spaces. Keeping the same style for the same terms aids readability.

Reply:
At the risk of continuing to bikeshed, the `H_J_c_T * rhs` needs the spacing by PEP8 as the `*` is the highest priority operator. It's not a consistency thing.

I  have no objections to not having the spaces for the other as it is correct. (I think it is a bit of an overreach to say PEP8 prefers it, much less recommends it.) I probably wasn't paying close enough attention here. (Although I do find it easier to read with all the underscores in the variable names, but its not important.)

Reply:
I've just added the spaces.